### ISSUE TITLE
Too-strict version restrictions on NumPy break distribution builds

### ISSUE THREAD

**Issue by ahesford** [scipy/scipy#18767]
I asked the same question [over at scikit-image](https://github.com/scikit-image/scikit-image/issues/7004), but I'm still not clear on where there is a hard requirement on a very specific version of NumPy [in the `pyproject.toml`](https://github.com/scipy/scipy/blob/v1.11.0/pyproject.toml#L43) for both of these packages. Having this specifc version of NumPy breaks Void Linux builds; patching out the version restriction doesn't seem to break builds.

If there are actual API changes that break SciPy builds, a range of supported versions seems more appropriate. If this restriction is instead used to enforce a specific version for isolated wheel builds, there must be a more appropriate way to add these restrictions to the venv.

**Comment by h-vetinari** [comment]
This reflects the conditions from https://github.com/scipy/oldest-supported-numpy/; depending on the oldest-available numpy is an important choice, because if one ends up building against (say) 1.25, then you will have a run-time dependence against numpy>=1.25. This would be way more aggressive than what we want to support.

It's therefore not a good idea to patch this out. It seems Void Linux has no 1.23.2 available. Do you have any other 1.23.x release? That would be essentially equivalent.

**Comment by ahesford** [comment]
I understand the need to support the oldest possible version *for the wheels that you distribute*. However, the `==` is still not the right way to handle this:
1. As you note, any other `1.23.x` version of NumPy should be sufficient, but any PEP517 builder will fail unless it finds the **exact** version specified. Even this requires patching of `pyproject.toml`.
2. In a source distribution (which is what Void uses to build its package), there is no need to compel building against the oldest supported NumPy version. Void is a rolling-release distribution and provides a fairly up-to-date version of NumPy---currently `1.25.0`. When we update SciPy, it is built with the currently packaged version of NumPy, and future updates of NumPy (should) retain backwards compatibility for SciPy at least for some deprecation period. There is no need for Void to support older versions because we only guarantee consistent functionality for all current package versions. (Generally, Void allows piecemeal package updates, but the guiding principle with partial updates is *caveat emptor*.)

I submit that the right approach here is to release source distributions with `pyproject.toml` restrictions that are as loose as possible. In general, this should be `numpy >= 1.23.2` for Python 3.11 (unless you *know* that some future release actively breaks buildability, in which case the addition of an upper bound would be appropriate). Anybody building from source should be able to build with whatever version of NumPy is *sufficiently new* rather than *exactly specified*. When the SciPy organization builds wheels for distribution, the virtual environment used to create those wheels should be responsible for installing the oldest supported NumPy to provide the wheel with the broadest compatibility.

**Comment by h-vetinari** [comment]
It's possible that you don't care if you build against newest numpy, but it's the right default _not_ to do that (and many other projects care about exactly this).

I empathise though, because I encounter this problem as well when building various packages for conda-forge. However, you It should be noted that scipy in conda-forge builds fine on 3.11 against numpy 1.23.5 despite that information in `pyproject.toml`.

We [use](https://github.com/conda-forge/scipy-feedstock/blob/main/recipe/build.sh):
```
# build
python -m build -w -n -x [...]  # -wnx flags mean: --wheel --no-isolation --skip-dependency-check
# install
pip install dist/scipy*.whl
```
It's not so interesting here why we need to separate build & installation, but if you need a way to opt-out of the dependency-checks, that's at least a possibility. It's probably also possible to pass that flag to `pip install` directly somehow?

CC @rgommers

**Comment by tylerjereddy** [comment]
> When the SciPy organization builds wheels for distribution, the virtual environment used to create those wheels should be responsible for installing the oldest supported NumPy to provide the wheel with the broadest compatibility.

`pyproject.toml` exists in part to avoid exactly that--it encodes our most portable wheel specification in a standardized way, so I'm pretty strongly against the suggestions to our distribution policy being made here. If you want to build a less-portable version of SciPy from source, with whatever dependencies you feel like, go ahead and disable the build isolation and make *your* own custom non-standard-compliant environment, rather than asking us to change our standard approach to distribution.

From the `pip install` docs, use this option:

> --no-build-isolation

> Disable isolation when building a modern source distribution. Build dependencies specified by PEP 518 must be already installed if this option is used.

https://peps.python.org/pep-0518/

Put whatever you want in *your* virtual environment.

It seems crazy to me that we'd even consider changing our `pyproject.toml` in some custom way when downstream consumers (including other distributions) can just use a simple command line option built into the ecosystem to do whatever they want (at their own peril, but the flexibility is yours). For those who trust our judgement, or don't even know what is going on under the hood, the current default to make the most portable wheel makes the most sense to me by quite a wide margin.

**Comment by ahesford** [comment]
We build PEP517 packages (including SciPy, as of today) with `python3 -m build --no-isolation --wheel`. I'm not quite sure how I feel about `--skip-dependency-check`. On the one hand, presumably any missing build dependencies will trigger a hard failure at *some* point in the build, so maybe the checks aren't necessary. On the other hand, there may be some legitimate restrictions that should be enforced by the dependency check.

@tylerjereddy you are miscontruing the purpose of PEP518. The purpose of `build-system.requires` in `pyproject.toml` is exactly what its name implies: a specification of the **requirements** to build a Python package, not a representation of **policy**. Encoding specific versions in this field to enforce a compatibility policy is both nonstandard and an abuse of the field. You're doing this to make up for a lack of accommodations for these kinds of policy preferences in PEP517/518. This is yet another example of Python packaging running amok without proper regard for distribution packaging. Their general attitude of "just live in our ecosystem" is hard to tolerate in a world that supports more than pure Python packages, where disribution packaging can be important to providing support for multiple architectures and C libraries and even optional dependencies in a sane manner.

I'm not asking you to consider changing `pyproject.toml` "in some custom way", and we already build with isolation disables. The issue here is that SciPy actually **breaks** the true intent of `build-system.requires`. I'm asking you to leave `requires` strictly for **requirements** (which, by your project's own admission in the comments about "Python versions which aren't officially supports", does **not** include a specific NumPy version) and find a more appropriate mechanism for enforcing policy preference in your wheel distribution workflow. Maybe the Python build ecosystem needs a "build-system.preferred-versions` key or something to automate this, but hijacking `build-system.requires` is not the answer.

As for users who "trust [your] [judgment] or don't even know what is going on", they should by using the wheels you distribute. By their very nature, *anybody* who is building a custom Python package from your source distribution knows what is going on and has a legitimate interest in not requiring an arbitrarily specific version of NumPy to do so.

**Comment by h-vetinari** [comment]
> I'm not quite sure how I feel about `--skip-dependency-check`.

It simply means you are responsible for ensuring that you've got the dependencies in place.

> The issue here is that SciPy actually **breaks** the true intent of `build-system.requires`.

This is IMO a philosophical debate reminiscent of the whole `requires-python` discussion. Feel free to propose `build-system.preferred-versions` on https://discuss.python.org/, but for the time being, the overriding principle is safe defaults in an ecosystem with inadequate tooling and metadata (and thus far from philosophical purity). You have a way to opt out if you know what you're doing.


> By their very nature, _anybody_ who is building a custom Python package from your source distribution knows what is going on and has a legitimate interest in not requiring an arbitrarily specific version of NumPy to do so.

I think this characterization is way too optimistic. Source installs still happen unintentionally in a wide variety of scenarios.

**Comment by eli-schwartz** [comment]
> It's possible that you don't care if you build against newest numpy, but it's the right default _not_ to do that (and many other projects care about exactly this).

No, absolutely not. :)

When locally building software for local-only use, the "correct" way to build software -- any software, it doesn't matter what, doesn't matter whether it's part of the python ecosystem or not -- is to build a local system specific binary edition of that software. Users doing this frequently have a habit of injecting compiler flags such as -march=native, and they are justified in doing so.

When building binaries for broad publication, the best thing to do is to build them according to some binary support matrix. This typically starts with encoding the minimum required libc, which is why things like the manylinux standard exist. At such time, you will *also* want to build against the oldest supported numpy, because arbitrary users downloading arbitrary versions of both numpy and scipy need to have that work given the binary distribution going on.

As for people building from source, scipy policy is pretty clear on that there are three types of users building from source:
- scipy developers, who know what they are doing
- distributions building scipy for their own curated ecosystem
- people who made a grave mistake and are probably going to regret trying to build from source, but "here are the very complicated docs on doing this correctly, prepare for troubleshooting"


As far as the pins go, these exist, I believe, solely to ensure that cibuildwheel behaves correctly. It's not completely bizarre for people building curated ecosystems to express the desire to have the pins not interact badly with their build pipelines. Whether that means scipy should change its pyproject.toml is not guaranteed, but it is ***not strange or incorrect*** for people to want that an ideal world would have this change.

**Comment by eli-schwartz** [comment]
> I think this characterization is way too optimistic. Source installs still happen unintentionally in a wide variety of scenarios.

I agree -- for example, 32-bit Windows.

But, anyone who triggers the building-from-source path without knowing, is already, most likely, going to get build errors for about a dozen different reasons.

The likelihood of their successfully building from source, and ending up with a locally built scipy wheel that works correctly ***until they downgrade numpy*** is... low. And even if they do manage to successfully build that wheel, it still works until and unless they locally downgrade numpy.

**Comment by eli-schwartz** [comment]
> Feel free to propose `build-system.preferred-versions` on https://discuss.python.org/, but for the time being, the overriding principle is safe defaults in an ecosystem with inadequate tooling and metadata (and thus far from philosophical purity). You have a way to opt out if you know what you're doing.

This, I believe, is the core of the issue -- and it should have been the first response, not "this is the ideal way to do things, you're holding your build pipelines wrong if you build against overly restrictive numpy versions".

**Comment by ahesford** [comment]
Whether or not PEP517 can be interpreted to endorse policy preferences in build requirements, I will note that, of the 213 Python packages in Void Linux that build according to PEP517, SciPy and scikit-image are the *only* two that do so.

Because PEP517 doesn't actually encode this use in the standard, we're forced to fall back on *de facto* standards. The practice of two (interconnected) projects in a sea of more than 200 cannot reasonably be considered to set the standard here.

**Comment by h-vetinari** [comment]
> It's not completely bizarre for people building curated ecosystems to express the desire to have the pins not interact badly with their build pipelines. Whether that means scipy should change its pyproject.toml is not guaranteed, but it is _**not strange or incorrect**_ for people to want that an ideal world would have this change.

Agreed on the desire to improve. In the current situation though, I think it's a pragmatic choice for projects to depend on `oldest-supported-numpy`, because _they have no control over how their project gets distributed_ (except for PyPI), and it's not reasonable to hope that all distributors big and small[^1] will do this correctly (hence, specifying it is a safer default).

[^1]: as small as some company-internal "approved" package rebuilds

> I will note that, of the 213 Python packages in Void Linux that build according to PEP517, SciPy and scikit-image are the _only_ two that do so.

I doubt that ~200 packages is representative. I get ~1700 [hits](https://github.com/search?q=%2Foldest-supported-numpy%2F&type=code) when searching github for "oldest-supported-numpy", and very likely nearly all of those use it for it's _only_ purpose, which is: the lowest-supported numpy at build-time. And that's not counting others ([example](https://github.com/cvxpy/cvxpy/blob/v1.3.2/pyproject.toml)) who do something equivalent themselves.

**Comment by ahesford** [comment]
The "equivalent" use in your example doesn't specify a specific version, it specifies a range that includes all possible patch versions, which is far more flexible than what SciPy does. (I still argue that it's inappropriate, but it's better.)

The first page of the linked search results are filled with a lot of comments and several uses where it is entirely appropriate (for example, in a script to build wheels, or in the construction of a rod environment for testing). Several other  hits in that list are just comments about what the package does.

Still more are lists in `requirements.txt` files which are frequently *advisory* rather than compulsory lists when the package is built. Most pre-PEP517 build scripts provide those requirements to aid in the construction of build or runtime environments but do not enforce their version restrictions (or, indeed, the availability of ANY version) during the build. While this is one of the things PEP517 was designed to address, it's telling that only one of the first-page hits on that search uses `oldest-supported-numpy` in a `pyproject.toml` or `setup.cfg`.

I'd also argue that Void provides a more representative sample than a search for a specific package designed to hard-code this kind of version restriction. You're biasing your sample by selecting only projects that refer to a package that does what you claim is appropriate, while my sample is just whatever Python packages people find useful on general Linux systems and have made an effort to package.

**Comment by tylerjereddy** [comment]
> @tylerjereddy you are miscontruing the purpose of PEP518. The purpose of build-system.requires in pyproject.toml is exactly what its name implies: a specification of the requirements to build a Python package, not a representation of policy. 

I'm going to have agree with @h-vetinari here, this sounds a bit philosophical, and we've had a fair number of long-winded issues discussing version pins already, that ultimately landed on some practical decision that has worked well for us, and can still be circumvented with straightforward command line flags for "advanced" uses.

Now, if there's a legitimate scenario where the command line flags are not sufficient to enable flexibility, I'd like to know about that, but as far as I can tell this issue is about "required" vs. "preferred/policy" metadata and that if we interpret those in some formal legal way then technically SciPy should rethink its approach to distributing so that some Linux distros don't need to add one or two command line flags?

At least from the standpoint of the SciPy release manager, I'm not seeing something that can't be worked around--if you really were blocked, I'd be more concerned, but if this is basically "I don't want to add a few command line flags and formally according to the PEP I shouldn't have to," that seems like a tougher sell?

Obviously, the SciPy release managers may not be particularly sympathetic, given the time and energy that has already gone into `pyproject.toml` pins working **in practice**, though Eli's point about just directing folks upstream seems sensible enough vs. suggesting that a less-portable build is "wrong."

**Comment by eli-schwartz** [comment]
In terms of practical value, I think the main reason a distro would prefer not to pass the option to skip dependency checks, is because it is a convenient way to programmatically check that the build recipe hasn't messed up when installing its own curated copies of the build dependencies.

This isn't strictly necessary as it can be manually verified, but it *is* a nice-to-have.

Absolute shower thought, but it would be neat if pip could by default install the build-requires using some kind of standardized constraints file. This probably has tie-in to the various failed proposals for a PEP-formalized lockfile format.

If such a thing did exist then you'd end up with a situation where building a wheel and installing its dependencies always used specific pinned versions, but building a wheel without build isolation would still permit using unpinned versions.

**Comment by ilayn** [comment]
A not directly relevant but very similar past discussion on why we are apparently using `Python-Requires` wrongly: #14738 which then led to https://discuss.python.org/t/requires-python-upper-limits/12663

In short, the word `require` in these PEP discussions has an additional invisible `-as-much-as-you-can-please` suffix attached to it which we didn't see. In turn, all kinds of pip machinery built on top of this "advisory" intent which we "broke" apparently. 

Most of these discussions assume that we are doing these just because of our stylistic choices and not having any practical reasons. I am a bit frustrated about that when this happens everytime. But it takes typically ~50 comments and some discuss.python entries only to be forgotten for another year when someone else realizes we are pinning things and should be stopped.

**Comment by h-vetinari** [comment]
> The "equivalent" use in your example doesn't specify a specific version, it specifies a range that includes all possible patch versions, which is far more flexible than what SciPy does. (I still argue that it's inappropriate, but it's better.)

I personally would be fine with using `>=1.23.2,<1.24` etc. also in `oldest-supported-numpy`, though I must note that I'm not aware why it was previously done like that (Chesterton's fence, etc.). That said, the relevant information here is always the feature level, which in our case is the minor version. Could you open an issue for that on https://github.com/scipy/oldest-supported-numpy?

**Comment by h-vetinari** [comment]
> If such a thing did exist then you'd end up with a situation where building a wheel and installing its dependencies always used specific pinned versions, but building a wheel without build isolation would still permit using unpinned versions.

Python packaging is full of such possible improvements, but actually shepherding them through design, discussion, standardization, roll-out, etc. until they can be considered "generally available" is a hardcore grind.

**Comment by rgommers** [comment]
I'll make a few points briefly, because I'm short on time:

- Yes, this is a real problem, and it's due to a gap in Python packaging design/standards, as has been pointed out. The main issue is https://discuss.python.org/t/relaxing-or-clarifying-pep-621-requirements-regarding-dynamic-dependencies/24752/14
- It is not just SciPy and scikit-image, it's any project that depends on NumPy's C API or on a C/C++ API from other projects. That is discussed in detail in https://pypackaging-native.github.io/key-issues/abi/ 
- What the `pyproject.toml` entries mean is ambiguous; we have no way to express both general/abstract project dependencies and wheel build requirements. See https://discuss.python.org/t/relaxing-or-clarifying-pep-621-requirements-regarding-dynamic-dependencies/24752/14. 
  - I think the rough consensus between Python packaging folks that understand the issue is that we prioritize setting the dependencies for wheel building for redistribution on PyPI.
  - That then indeed is a problem for Linux distros, which can and should be worked around by disabling build isolation and skipping `pypa/build`'s dependency check (which arguably should be automatically disabled when disabling build isolation, but that's another story).

Most important point: **NumPy 1.25.0 changes how all this works, so for the next release we will drop these pins** (and hence we will not make minor tweaks now, that's pretty much pointless). See https://numpy.org/doc/1.25/dev/depending_on_numpy.html#adding-a-dependency-on-numpy for context.

> But, anyone who triggers the building-from-source path without knowing, is already, most likely, going to get build errors for about a dozen different reasons.

This isn't quite true on Linux, things tend to work reasonably well - users typically have compilers installed, so if one also has BLAS/LAPACK, the build typically works. And users on platforms like `ppc64le` or Alpine, for which we don't supply wheels, kinda rely on this stuff.

**Comment by ahesford** [comment]
I appreciate the discussion, which was always philosophical as I've already worked around the issue for Void. My intent was to seek clarification on the purpose of these pins and, once their role in wheel building was explained, to illuminate the conflicting constraints of wheel distribution and local building.

As eli-schwartz notes, disabling dependency checks is not desirable because they help us catch oversights in dependencies that sometimes break obviously but can often lead to subtle issues that are hard to track. For example, in the easy-install days, we had several Void packages that just silently grabbed dependencies over the Internet instead of building against our packaged versions simply because package authors forgot to include one or more dependencies in our build templates. For our purposes, patching out the version requirement is preferable. I look forward to the day when NumPy 1.25 will make this unnecessary.

The priority on PyPI distribution of wheels is consistent within PyPA and is not entirely unreasonable, although it can be frustrating for Linux distributors who are unable to just use pip and move on. Not only do we have to contend incorporating Python packages into our package graphs, but some CPython native extensions simply must be built against system libraries and therefore must contend with non-isolated builds. (For example, I had a private project that used ZeroMQ in a C library and a Python helper that would manipulate the sockets and call the library via a CFFI or pybind11 wrapper. The version of PyZMQ I used absolutely *had* to be linked with my system ZeroMQ libraries... any prebuilt wheel would return context and socket handles that were completely independent from those in the C library.)

I am aware that PyPA and others involved in the debates that advance expectations and standards for Python packaging perceive a lack of involvement in the process by Linux distributors. While I can't speak for others, I admit that I am generally aloof until something breaks my workflow. In my defense, the Void Linux team is small and our attention is spread widely; it is difficult to notice and engage in these discussions early enough to avert inconvenience.

**Comment by rgommers** [comment]
> Not only do we have to contend incorporating Python packages into our package graphs, but some CPython native extensions simply must be built against system libraries and therefore must contend with non-isolated builds

I think every distro should always disable build isolation for any usage of `pip` or `build` - and not some but all extension modules must be built against system libraries. Pip has an env var for that; conda-forge for example sets it by default so it's impossible to use build isolation by accident.

**Comment by ahesford** [comment]
> > Not only do we have to contend incorporating Python packages into our package graphs, but some CPython native extensions simply must be built against system libraries and therefore must contend with non-isolated builds
> 
> I think every distro should always disable build isolation for any usage of `pip` or `build` - and not some but all extension modules must be built against system libraries. Pip has an env var for that; conda-forge for example sets it by default so it's impossible to use build isolation by accident.

Yes, we build everything without isolation. The issue I raised here is that use of the build requirements to enforce a specific NumPy version that guarantees wide compatibility for distributed wheels essentially presupposes the use of build isolation (which is why `==1.23.2` doesn't cause alarm for the SciPy development team; it will always just be available!). When we raise these kinds of PyPA issues, we often face backlash that amounts to "you should just be building in isolation", which is clearly untenable.

**Comment by ahesford** [comment]
> It doesn't. Our actual build-time requirement on `numpy` is the same as our runtime requirement, so for `1.11.0` it's:
> 
> ```
> numpy>=1.21.6,<1.28.0
> ```
> 
> We just don't have a way to express that in metadata.

But this is my central argument: that `build-system.requires` is the place where `numpy>=1.21.6,<1.28.0` should be listed to express the *actual* requirements, while the desire to build distributable wheels against `numpy==1.23.2` for broad and supported compatibility belongs... somewhere more appropriate.

**Comment by rgommers** [comment]
Yeah .... that's the core of this mess indeed. I wish I could fully agree with you here, because the actual generic project requirements are more important imho than the wheel-for-pypi-specific ones. But I unfortunately can't given the state of Python packaging, it needs work. As it stands now, `pyproject.toml` is for the wheel-specific ones:(

**Comment by tylerjereddy** [comment]
Things have changed a bit for the `1.13.0` series with NumPy 2.0.x as wheel build target, and better backwards compatibility guarantees from that, but this is a long/tricky discussion so I'll bump the milestone.

The version of `pyproject.toml` with Ralf's updates on the release branch is here: https://github.com/scipy/scipy/blob/maintenance/1.13.x/pyproject.toml. That does look more flexible than the hard `==`, but I won't try to close this issue on the basis of that alone. Ralf did also add a comment with some advice for distributors in there.

**Comment by rgommers** [comment]
I just noticed that that 1.13.x change didn't land in `main` yet. I'll open a new PR that I think will close this issue.
